### PR TITLE
Improved stream resolution to allow collapsed (Nd)Overlays

### PR DESCRIPTION
**Pull request by philippjfr**
The code that resolves which plot to attach a linked stream to previously gave up when some part of the graph reduces an (Nd)Overlay by one or more layers since it cannot easily determine which layers were dropped. However in the case of operations that collapse multiple input layers into a single layer, e.g. the new datashader stack operation which combines multiple RGBs into one, there is no ambiguity since information from all the input layers persists. This means that an operation like this:

```python
stream = [RangeXY()]
stack(datashade(points1, streams=[stream]) * datashade(points2, streams=[stream]))
```

now correctly resolves the linked stream and therefore responds to zoom events. In order to provide more flexibility I've also exposed the ``link_inputs`` argument on the ``DynamicMap.map`` method in case linking is not desired.

Here's an example:

collapse_streams

```
   Points              Points
     |                   |
 datashade           datashade
     |                   |
    RGB ----- * ------- RGB
              |
           Overlay          <-- Previously gave up resolving streams here
              |      ^
            stack    |
              |      |  Direction of resolution
             RGB     |
```

### CODE REVIEW

**Comment by jbednar** [comment]
Great, thanks!

**Comment by philippjfr** [comment]
Ready to merge.

File: holoviews/plotting/util.py [hunk 1]
```diff
@@ -130,23 +130,26 @@ def compute_overlayable_zorders(obj, path=[]):
     # Process the inputs of the DynamicMap callback
     dmap_inputs = obj.callback.inputs if obj.callback.link_inputs else []
     for z, inp in enumerate(dmap_inputs):
+        no_increment = False
```
Comment (jlstevens) [inline holoviews/plotting/util.py]: Should this be called ``no_z_increment``?

Comment (jlstevens) [reply holoviews/plotting/util.py]: Or ``zorder_unchanged``?

Comment (philippjfr) [reply holoviews/plotting/util.py]: Yes, sure.

File: holoviews/core/spaces.py [hunk 2]
```diff
@@ -1062,19 +1062,22 @@ def _cache(self, key, val):
         self[key] = val
 
 
-    def map(self, map_fn, specs=None, clone=True):
+    def map(self, map_fn, specs=None, clone=True, link_inputs=True):
```
Comment (jlstevens) [inline holoviews/core/spaces.py]: Seems useful though I just want to check this feature is unrelated to the rest of the PR...

**Comment by jlstevens** [comment]
Happy to merge once my one query and naming suggestion in the comments above are addressed.

File: holoviews/core/spaces.py (hunk 2 above)
Comment (philippjfr) [reply holoviews/core/spaces.py]: It's related but not directly I guess. Basically if we are going to allow operations which collapse layers to still link the inputs to that operation we should also make sure that the user can disable that behavior.

Comment (jlstevens) [reply holoviews/core/spaces.py]: That makes sense. Thanks.

**Comment by jlstevens** [comment]
Looks good and the tests are passing.
